src/main/java/com/vmware/vipclient/i18n/messages/service/ComponentService.java

```diff
@@ -108,38 +106,48 @@ public MessageCacheItem getMessages(Iterator<Locale> fallbackLocalesIter) {
     		}
     	} else { // Item is not in cache.
 			ProductService ps = new ProductService(dto);
-			if (!ps.isSupportedLocale(Locale.forLanguageTag(dto.getLocale()))) { // Requested locale is not supported
-				Locale matchedLocale = LocaleUtility.pickupLocaleFromList(new LinkedList<>(ps.getSupportedLocales()), Locale.forLanguageTag(dto.getLocale()));
-				if (ps.isSupportedLocale(matchedLocale)) { // Requested locale matches a supported locale (eg. requested locale "fr_CA matches supported locale "fr")
+			Locale locale = Locale.forLanguageTag(dto.getLocale());
+			if (ps.isSupportedLocale(locale) || VIPCfg.getInstance().isPseudo()) {
+				cacheItem = createCacheItem(); // Fetch for the requested locale from data store, create cacheItem and store in cache
+				if (cacheItem.getCachedData().isEmpty())  // Failed to fetch messages for the requested locale
+					cacheItem = getMessagesFallbackLocale(fallbackLocalesIter);
+			} else { // Requested locale is not supported
+				Locale matchedLocale = LocaleUtility.pickupLocaleFromList(ps.getSupportedLocales(), locale);
+				if (matchedLocale != null) { // Requested locale matches a supported locale (eg. requested locale "fr_CA matches supported locale "fr")
 					MessagesDTO matchedLocaleDTO = new MessagesDTO(dto.getComponent(), matchedLocale.toLanguageTag(), dto.getProductID(), dto.getVersion());
-					return new ComponentService(matchedLocaleDTO).getMessages();
-				}
+					cacheItem = new ComponentService(matchedLocaleDTO).getMessages();
+				} else  // Requested locale is not supported and does not match any supported locales
+					cacheItem = getMessagesFallbackLocale(fallbackLocalesIter);
```

Comment:
We can unify this these 2 situations. If matchedLocale == null, then set locale to fallback locale. Get its translation afterwards.

Reply:
You have a similar comment as Huihui. Pasting my reply below:

The problem with this is due to the legacy code where locale is part of the DTO. If the matched locale is not null, there are 2 possible scenarios:

a. The requested locale is a supported locale (exact match) -- lines 110-113
Here, createCacheItem in line 111 fetches the matched locale and caches it. Take note that it uses the dto's locale to generate the cache key.
b. The requested locale is not an exact match (e.g. fr-CA which is not supported matches fr which is supported) -- lines 117-118
Here, createCacheItem must be done on the supported locale (fr), not the requested locale (fr-CA). If I directly call createCacheItem for this case, the cache key will have fr-CA for the same set of fr messages, causing duplicate sets of messages in the cache, which is not good. So, call ComponentService(matchedLocaleDTO).getMessages instead.

This is why these 2 scenarios are handled differently. We can refactor this in the future and extract locale out of the DTO, but I did not want to touch that legacy code for now.

Reply:
I think that our comments are different. I'm saying 115~120.

Reply:
I do not understand then. Could you please provide a code suggestion?

Reply:
Please check this code suggestion for 115 - 120:

ProductService ps = new ProductService(dto);
			Locale localeToQuery = Locale.forLanguageTag(dto.getLocale());
			if (!ps.isSupportedLocale(localeToQuery)) {
				localeToQuery = LocaleUtility.pickupLocaleFromList(ps.getSupportedLocales(), localeToQuery);
				if (localeToQuery == null) {// Requested locale is not supported and does not match any supported locales
					if (!dto.getLocale().equals(ConstantsKeys.SOURCE) && fallbackLocalesIter != null && fallbackLocalesIter.hasNext()) {
						localeToQuery = fallbackLocalesIter.next();
					}
				} else {
					return new MessageCacheItem();
				}
			}

			MessagesDTO matchedLocaleDTO = new MessagesDTO(dto.getComponent(), localeToQuery.toLanguageTag(), dto.getProductID(), dto.getVersion());
			cacheItem = new ComponentService(matchedLocaleDTO).getMessages();
    	}

Reply:
Thanks for the code suggestion. It will work, but I think what I currently have is cleaner and has less lines of code, don't you think? Basically, your `if (!dto.getLocale().equals(ConstantsKeys.SOURCE) && fallbackLocalesIter != null && fallbackLocalesIter.hasNext()) `

is a repeat of line 132.